### ISSUE TITLE
Assistance Request: Push Many Large Files Over Flaky Internet Connection

### ISSUE THREAD

**Issue by lbell** [odeke-em/drive#825]
Hoping somebody might know how to best implement this program (likely via cron) to keep trying to push a large directory (hundreds of thousands of 4-10mb files) over a flaky (and slow*) internet connection until it succeeds.

- Would starting a new instance every 10 minutes and hoping for the best be a bad approach (ie, if I start the push on startup, will the program keep trying until finished or does it abort with a dropped internet connection?)
- Does this program have a graceful "resume" feature if the computer is shut down before the push is finished?
- As files are added to the directory, will it put them into the queue if a push has already started?

*And when I say flaky / slow, I'm talking third-world country speeds that rarely reach 3G on a good day, and reliably drop out every 20-40 minutes for a few minutes...

**Comment by chevillet** [comment]
Unfortunately, I'm afraid that drive will not work for you.

I had a similar situation once (super slow connection with plenty of random interruptions) and couldn't even figure out a way to do this with any Google Drive "clients".
I went for Dropbox because the official client handles poor internet connections very well (resume, conflicts and so on).

So if someone has a solution for Google Drive, I'd be interested too.

**Comment by lbell** [comment]
Yes, Dropbox works amazingly well in this situation... unfortunately we have a free unlimited drive account, and can't afford a Dropbox subscription.

**Comment by chevillet** [comment]
OK, I see. I hope you'll find a solution and I'll also be monitoring this issue closely.

**Comment by odeke-em** [comment]
Hello there @lbell, thank you for the question and welcome to drive!

Thank you @chevillet for helping out with the answers!

@lbell, so in regards to your questions:  
`Would starting a new instance every 10 minutes and hoping for the best be a bad approach (ie, if I start the push on startup, will the program keep trying until finished or does it abort with a dropped internet connection?)`
If the connection to Google Drive's backend is flaky, the default transport will try to reconnect until it gives up. However, drive will retry the entire operation, pushing from the first byte.
It retries with exponential backoff for failed connections. You can investigate this by using option `--verbose` when doing a push or pull e.g
```shell
$ drive push --verbose <your_paths...>
```

`Does this program have a graceful "resume" feature if the computer is shut down before the push is finished?`
Unfortunately, it doesn't. Pushing files to Google Drive for now is just an all or nothing operation. We use resumable media uploads in the transport but that isn't exposed to drive itself so we can't save the pushed progress and retry from there. If that we possible, we'd be able to implement partial updates instead of reuploading the entire file.


`As files are added to the directory, will it put them into the queue if a push has already started?`
No, it won't. We don't have a directory watcher implemented. Only the files in the manifest at push time will be pushed up. On next push it will pick up the unpushed files.

I agree with what @chevillet said, Google Drive's backend and by extension drive seems to be made for places with reliable connections or places where uploading huge chunks is not a big deal :(
Perhaps such needs might get us innovating on actually implementing proper cloud storage.
